src/box/vy_scheduler.c

```diff
@@ -1116,6 +1117,131 @@ vy_task_write_run(struct vy_task *task, bool no_compression)
+ * t_r - Adopted run_size_ratio_coeff.
+ *
+ * Note:
+ * We should consider the fact that
+ * current range doesn't yet contains newly created run.
+ *
+ * @param lsm
+ * @param range
+ * @param new_run_stmts_count statements count in the newly created after
+ * dump/compaction run
+ * @param W total number of runs in the range
+ * after operation (dump/compaction)
+ * @param total_stmts_count total number of statements in the range
+ * after operation (dump/compaction)
+ * @param skip_n number of runs to skip during the range traverse
+ * as they are compacted into one new run.
+ * @return optimally calculated bloom_fpr
+ */
+double
+vy_task_find_range_bloom_fpr(
+	struct vy_lsm *lsm,
+	struct vy_range *range,
+	uint64_t new_run_stmts_count,
+	int W,
+	uint64_t total_stmts_count,
+	int skip_n)
+{
+	double R = W * lsm->opts.lookup_cost_coeff;
+
+	if (new_run_stmts_count == total_stmts_count) {
+		/** By definition of R. May also look at formulas ahead.
+		 *  R <= 1 in this case as soon as W = 1.
+		 */
+		return R;
+	}
+
+	struct vy_slice *last_slice =
+		rlist_last_entry(&range->slices, struct vy_slice, in_range);
+	uint64_t last_slice_stmts_count = last_slice->count.rows;
+	uint64_t scnd_last_slice_stmts_count = 0;
+
+	if (skip_n == range->slice_count - 1) {
+		/** In both dump and compaction cases it means
+		 *  we only have the new run except the last run.
+		 */
+		scnd_last_slice_stmts_count = new_run_stmts_count;
+		if (scnd_last_slice_stmts_count > last_slice_stmts_count) {
+			/** In case, for example, when the size of dumped run
+			 *  is bigger than the size of the last run.
+			 */
+			return lsm->opts.bloom_fpr;
+		}
+	} else {
+		struct vy_slice *scnd_last_slice = rlist_prev_entry(
+			last_slice, in_range);
+		scnd_last_slice_stmts_count = scnd_last_slice->count.rows;
+	}
+
+	int W_u = 0;
+	int W_f = 0;
+	double t_r = 0;
+
+	if (R <= (total_stmts_count / last_slice_stmts_count)) {
+		W_u = 0;
+	} else {
+		t_r = (double)scnd_last_slice_stmts_count /
+		      last_slice_stmts_count;
+		W_u = MAX(0, floor((R - t_r * W) / (1 - t_r)));
+		if (W_u == W) {
+			return 1;
+		}
+	}
+	W_f = W - W_u;
+
+	int slice_number = 0;
+	uint64_t filtered_stmts_count_sum = new_run_stmts_count;
+	struct vy_slice *slice;
+	rlist_foreach_entry(slice, &range->slices, in_range) {
+		if (slice_number++ < skip_n) {
+			continue;
+		}
+
+		filtered_stmts_count_sum += slice->count.rows;
+
+		if (slice_number == W_f) {
+			break;
+		}
+	}
+
+	double bloom_fpr = (R - W_u) * new_run_stmts_count /
+			   filtered_stmts_count_sum;
+	return bloom_fpr;
+}
+
+/*
+ * Logic:
+ * Let's consider our dump is divided evenly into all lsm ranges.
+ * So everything we have to do is to find optimized bloom_fpr
+ * for small dump part in one of lsm ranges.
+ *
+ * @param lsm
+ * @param new_run_stmts_count statements count in the newly created after
+ * dump run
+ * @return optimally calculated bloom_fpr for dump
+ */
+double
+vy_task_find_lsm_bloom_fpr(
```

Comment:
I don't understand the name of this function.

Reply:
I would like to call it `vy_task_find_bloom_fpr_for_run_in_range`, but the name seems too long for me. Can't come up with better name.
As a variant I can move this function to vy_run.c file and name it like `vy_run_find_bloom_fpr_in_range`, but I don't think that this logic should lie in vy_run.c file.

Reply:
Let's move it to `box/vy_stmt.{h, c}` and call it `vy_bloom_optimal_fpr_for_dump`.

Reply:
I think that vy_stmt.h may be not the best location because adding vy_lsm.h header in it causes an error `implicit declaration of function ‘vy_entry_compare’`. Static function `vy_entry_compare` used in vy_stmt.h -> vy_lsm.h -> vy_cache.h before declaration.